src/dump.c

```diff
@@ -2014,9 +2014,13 @@ DLLEXPORT jl_value_t *jl_uncompress_ast(jl_lambda_info_t *li, jl_value_t *data)
 
 DLLEXPORT int jl_save_incremental(const char *fname, jl_array_t *worklist)
 {
+    // make temporary file name foo.ji.XXXXXX
+    // (would be better to use mkstemp once this is in libuv -- see libuv/libuv#322)
+    char *tmpfname = mktemp(strcat(strcpy((char *) malloc(strlen(fname)+8), fname), ".XXXXXX"));
```

Comment:
I would think some sort of locking is really needed, whether via `flock`, or workers coordinating in some other fashion.  (I was used to using the built-in database to do this in similar situation, caching of JIT compilation of dynamic language, which had built-in locking and atomic operations, so it was easy)